### DISCUSSION TITLE
Inquiry into Custom Integration with TicketManager

### DISCUSSION THREAD

**Discussion by HoshiKurama** [plan-player-analytics/Plan#2431] [APIs and Development]
Hello! I’m the creator of TicketManager, a support-ticket tool. My users expressed a desire to fully manage tickets from Player Analytics, so I had a few questions if you didn’t mind.

- Developers develop their own pages, correct? I believed so at first, but after seeing the plugin request section, I’m not sure.
- Do you have a template for the components that must remain consistent? I figure all custom pages must keep the themes and navigational components consistent.
- Does Plan provide a means to retrieve the Minecraft UUID of the current user?
- The page will need access to the TicketManager database. How does Plan handle waiting for database query results?

Thank you very much in advance!

**Comment by AuroraLS3** [comment]
> Developers develop their own pages, correct? I believed so at first, but after seeing the plugin request section, I’m not sure.

This is possible yes. There is an API for adding HTTP request resolvers into the Plan Webserver (https://github.com/plan-player-analytics/Plan/wiki/APIv5-PageExtension-API) that the plugin also uses internally for its website.

The plugin request section is mostly filled with requests from non-developers who want to see some data from other plugins, usually through DataExtension API (More limited to just showing things).

> Do you have a template for the components that must remain consistent? I figure all custom pages must keep the themes and navigational components consistent.

Currently the plugin frontend is being rewritten in React so consistency with old version may look slightly different (https://github.com/plan-player-analytics/Plan/issues/2260) - but error.html is probably a good start regardless https://github.com/plan-player-analytics/Plan/blob/master/Plan/common/src/main/resources/assets/plan/web/error.html

It has a simple card with a title and paragraph that you can replace with whatever you want

> Does Plan provide a means to retrieve the Minecraft UUID of the current user?

PageExtension API `Request#getUser()#getUUID` https://github.com/plan-player-analytics/Plan/wiki/APIv5-PageExtension-API#request-object - Note that both getUser and getUUID return Optional, since User may not be present if authentication is disabled, and UUID may not be present if the user was registered via server console (and so not linked to any particular player)

> The page will need access to the TicketManager database. How does Plan handle waiting for database query results?

Any methods in the APIs are called from non-server threads so blocking to wait for database queries is fine inside any method.

----

There's probably some challenges so I'll put down some advice on how I'd approach this:

- So that users can see something, Implement a new `Resolver` that returns your custom html file (Possibly using error.html as starting point or using something else) (eg `/ticketmanager?player=uuid`)
  
  ```java
  public class YourResolver implements Resolver {
    @Override
    Optional<Response> resolve(Request request) {
        String html; // a html page (utf-8 string), read from your plugin jar or use ResourceService
        Response response = Response.builder()
            .setMimeType(MimeType.HTML)
            .setContent(html)
            .build();
        return Optional.of(response);
      }
  }
  ```
  (How to register new resolvers: https://github.com/plan-player-analytics/Plan/wiki/APIv5-PageExtension-API#registering-your-resolvers)
- To get your data make a HTTP request in javascript inside the html file you made https://javascript.info/fetch
- To have endpoint for that request, implement a 2nd Resolver that returns your data as json
  - You can probably have player UUID as a parameter in the URL, eg `/ticketmanager/data?player=uuid`
  ```java
  public class YourResolver implements Resolver {
    @Override
    Optional<Response> resolve(Request request) {
        Optional<String> playerUUID = request.getQuery().get("player");
        // Getting the data you need, sending back error if necesary https://plan-player-analytics.github.io/Plan/api/com/djrapitops/plan/delivery/web/resolver/exception/package-summary.html
        return Optional.of(
            Response.builder()
              .setMimeType(MimeType.JSON)
              .setStatus(200)
              .setJSONContent(Collections.singletonMap("example", "value"))
              .build();
        );
      }
  }
  // Later in some other part of code
  ResolverService.getInstance().registerResolver("TicketManager", "/ticketmanager/data", new YourResolver())
  ```
- So that users can navigate to the new part of the website, implement a DataExtension (https://github.com/plan-player-analytics/Plan/wiki/DataExtension-API---Getting-started) that has a link to the new page eg. (This will show up on the Players' pages in the Plan website)
```java
@PluginInfo(...)
public class TicketMasterExtension implements DataExtension {

    @StringProvider(...)
    public String link(UUID playerUUID) {
        return "<a href=\"../../ticketmanager?player=" + playerUUID + "\">Tickets from this player</a>"
    }
}
```

----

If something isn't clear please let me know and I'll try to clarify.

There's also Javadocs if that helps https://plan-player-analytics.github.io/Plan/

**Reply by HoshiKurama** [reply]
Thank you very much! This is all very helpful information. I'll certainly ask if I have more questions.
